client/landing/gutenboarding/hooks/use-on-unmount.ts

```diff
@@ -0,0 +1,28 @@
+/**
+ * External dependencies
+ */
+import { useEffect, DependencyList } from 'react';
+
+/**
+ * The user can either leave the page by navigating within the react app, or with browser navigation
+ * The onUnmount event will be triggered once in either case
+ * https://angelos.dev/2019/05/custom-react-hook-to-prevent-window-unload/
+ *
+ * @param onUnmount The function to be called when the component is unmounted
+ * @param deps The react dependencies to be passed in to useEffect
+ **/
+export const useOnUnmount = ( onUnmount: () => void, deps?: DependencyList ) => {
+	const onUnload = () => {
+		onUnmount();
+	};
+	useEffect( () => {
+		// Handles browser navigation events
+		window.addEventListener( 'beforeunload', onUnload );
```

Comment:
I haven't manually tested yet, but this will fire between Gutenboarding routes as well, i.e., if a user navigates within the app from, say, Intent Gathering to Design and back again?
 
Or do we still need to add a change listener to route changes?

Reply:
yea the beforeunload event isn't triggered when you navigate between routes in the same app, only if you leave the page. so that's why I also trigger the event in the cleanup function returned to useEffect.

Reply:
> so that's why I also trigger the event in the cleanup function returned to useEffect.

**:TIL:** 👍

Reply:
Nice, I was wondering how it worked!